### PR TITLE
[SPARK-24806][SQL] Brush up generated code so that JDK compilers can handle it

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr brushed up code so that JDK compilers can handle it because the master sometimes generates Java code that only janino can compile it. This is the issue I found when working on SPARK-24498.

## How was this patch tested?
Existing tests.

### CODE REVIEW

**Comment by maropu** [comment]
As I said in `SPARK-24498`, I think, whether `SPARK-24498` resolved or not, we'd be better to make generated code Java-compatible as far as possible.

As @mgaido91 suggested there, there is no test to check this now. I think it'd be nice if we could. But, since the compilation of JDK compilers is too slow (see the performance values I showed in `SPARK-24498`), I think it is impractical that jenkins checks if all the generated code can be compiled in JDK Java compilers...

**Comment by gatorsmile** [comment]
cc @rednaxelafx

**Comment by maropu** [comment]
Also, `SparkException` needs to extend `RuntimeException` instead of `Exception` because some generated codes do not have code to catch the exception. But, the change causes Mima test failures, so it is removed from this pr now.

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
I'll fix the failures soon.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 1]
```diff
@@ -3758,7 +3758,10 @@ case class ArrayUnion(left: Expression, right: Expression) extends ArraySetLike
       } else {
         val arrayUnion = classOf[ArrayUnion].getName
         val et = ctx.addReferenceObj("elementTypeUnion", elementType)
-        val order = ctx.addReferenceObj("orderingUnion", ordering)
+        // Some data types (e.g., `BinaryType`) have anonymous classes for ordering and
+        // `getCanonicalName` returns null in these classes. Therefore, we need to
+        // explicitly set `className` here.
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: nit: as we are adding this comment, shall we also mention that Janino works anyway, but JDK complains here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 2]
```diff
@@ -1585,6 +1585,9 @@ case class InitializeJavaBean(beanInstance: Expression, setters: Map[String, Exp
   }
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    // Resolves setters before compilation
+    require(resolvedSetters.nonEmpty)
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: why do we need to add this?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala [hunk 3]
```diff
@@ -318,7 +318,8 @@ case class SampleExec(
         v => s"""
           | $v = new $samplerClass<UnsafeRow>($lowerBound, $upperBound, false);
           | $v.setSeed(${seed}L + partitionIndex);
-         """.stripMargin.trim)
+         """.stripMargin.trim,
+        forceInline = true)
```
Comment (mgaido91) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: why do we need this?

**Review by mgaido91 — COMMENTED** [review]
thanks @maropu. As I already mentioned and you remembered here, my only concern here is the lack of testing. I am fine having this in, but without proper tests, I am afraid it could be not so useful (as we may missing other problems or other problems may be introduced later).

Anyway, as it is of no harm having it, if others agree that it is useful, I am fine having it in.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 3 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Generic arrays not allowed;
```
/* 019 */   private org.apache.spark.util.random.BernoulliCellSampler<UnsafeRow>[] sample_mutableStateArray_0 = new org.apache.spark.util.random.BernoulliCellSampler<UnsafeRow>[1];
--- 
 Cause: java.util.concurrent.ExecutionException: org.codehaus.commons.compiler.CompileException: failed to compile:
(Line 40, Column 101) generic array creation
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 4]
```diff
@@ -3758,7 +3758,14 @@ case class ArrayUnion(left: Expression, right: Expression) extends ArraySetLike
...
         val et = ctx.addReferenceObj("elementTypeUnion", elementType)
-        val order = ctx.addReferenceObj("orderingUnion", ordering)
+        // Some data types (e.g., `BinaryType`) have anonymous classes for ordering and
+        // `addReferenceObj` generates code below for anonymous classes;
+        //
+        //   (org.apache.spark.sql.types.BinaryType$$anon$1) references[0];
+        //
+        // Janino can compile this statement, but JDK Java compilers can't. So, we explicitly
+        // set `className` here.
+        val order = ctx.addReferenceObj("orderingUnion", ordering, "scala.math.Ordering<Object>")
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: @mgaido91 I a little misunderstood, so I updated the comment. I don't know why this fails in JDK compilers now, so I think I need to dig into this more. cc: @rednaxelafx

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 5]
```diff
@@ -890,7 +890,7 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
     case StringType =>
       val intOpt = ctx.freshName("intOpt")
       (c, evPrim, evNull) => s"""
-        scala.Option<Integer> $intOpt =
+        scala.Option<Object> $intOpt =
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: Out of curiosity, why is this change needed? Is it because it's really a scala int that's returned?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 6]
```diff
@@ -1588,6 +1588,15 @@ object CodeGenerator extends Logging {
 
   def primitiveTypeName(dt: DataType): String = primitiveTypeName(javaType(dt))
 
+  def getClassName(cls: Class[_]): String = {
+    try {
+      return Option(cls.getCanonicalName).getOrElse(cls.getName)
+    } catch {
+      case err: InternalError =>
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: What would cause InternalError here? just wondering if there's a cleaner way.
Will `getCanonicalName` ever return null? otherwise seems like you don't need to fall back to `getName` above. Also `return` isn't needed below?

**Comment by srowen** [comment]
PS I don't think SparkException should be a RuntimeException even if it were possible. It's possible to get the Scala code to declare SparkException in the byte code if that's what you need it to do for the benefit of a Java compiler -- `@throws[...]`

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
@srowen Thx for the comment. aha, I see, It looks reasonable to me. We should try to fix so.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 5 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: I don't look into the bytecode that JDK compilers generate though, the type of the return value seems to be erased in the bytecode;
```
...
/* 045 */       scala.Option<Integer> intOpt_0 =
/* 046 */       org.apache.spark.sql.catalyst.util.DateTimeUtils.stringToDate(((UTF8String) references[0] /* literal */));
/* 047 */       if (intOpt_0.isDefined()) {
/* 048 */         value_1 = ((Integer) intOpt_0.get()).intValue();
/* 049 */       } else {
/* 050 */         isNull_1 = true;
/* 051 */       }
/* 052 */
...
- Day / DayOfMonth *** FAILED ***
  Code generation of dayofmonth(cast(2000-02-29 as date)) failed:
  java.util.concurrent.ExecutionException: org.codehaus.commons.compiler.CompileException: failed to compile:
  (Line 67, Column 62) incompatible types: scala.Option<java.lang.Object> cannot be converted to scala.Option<java.lang.Integer>
  java.util.concurrent.ExecutionException: org.codehaus.commons.compiler.CompileException: failed to compile:
  (Line 67, Column 62) incompatible types: scala.Option<java.lang.Object> cannot be converted to scala.Option<java.lang.Integer>
        at com.google.common.util.concurrent.AbstractFuture$Sync.getValue(AbstractFuture.java:306)
        at com.google.common.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:293)
        at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:116)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 6 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Please see the commit https://github.com/apache/spark/commit/cc88d7fad16e8b5cbf7b6b9bfe412908782b4a45 and this is the same fix with [Utils.getSimpleName](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/util/Utils.scala#L2732). Yea, I see and I think this is workaround now.

Yea, in some cases, the doc says that `getCanonicalName` could return null;
```
    /**
     * Returns the canonical name of the underlying class as
     * defined by the Java Language Specification.  Returns null if
     * the underlying class does not have a canonical name (i.e., if
     * it is a local or anonymous class or an array whose component
     * type does not have a canonical name).
     * @return the canonical name of the underlying class if it exists, and
     * {@code null} otherwise.
     * @since 1.5
     */
    public String getCanonicalName() {
```

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Anyway, `return` removed.

**Comment by viirya** [comment]
Should this change be a part of #21777? Seems it shall be?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 3 above)
Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Though this may cause problems leading to the 64KB limit issue. So if we are not including the jdk support I'd be against this change...

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: aha, ok. let's us wait for the other developer's comments.

**Comment by maropu** [comment]
yea, if we get the consensus to implement #21777, it sounds ok to me.

**Comment by maropu** [comment]
retest this please
